src/util/nchan_accumulator.c

```diff
@@ -0,0 +1,198 @@
+#include <math.h>
+#include <nchan_module.h>
+
+#include "nchan_accumulator.h"
+
+// based on work done for Queris (c) 2011-2017 Leo Ponomarev
+// https://github.com/slact/queris/blob/master/lib/queris/indices.rb#L661-L712
+// licensed under the MIT license, same as Nchan
+
+int nchan_accumulator_init(nchan_accumulator_t *acc, nchan_accumulator_type_t type, double halflife) {
+  switch(type) {
+    case ACCUMULATOR_EXP_DECAY_FLOAT:
+      acc->data.ed_float.value = 0;
+      acc->data.ed_float.weight = 0;
+      acc->data.ed_float.lambda = (double)1 / halflife;
+      break;
+    
+    case ACCUMULATOR_SUM:
+      acc->data.sum.value = 0;
+      acc->data.sum.weight = 0;
+      break;
+      
+    default:
+      return 0;
+  }
+  ngx_memzero(&acc->last_update, sizeof(acc->last_update));
+  acc->type = type;
+  
+  return 1;
+}
+
+static double ngx_time_t_diff(ngx_time_t *t0, ngx_time_t *t1) {
+  if(t1->sec == t0->sec && t1->msec == t0->msec) {
+    return 0;
+  }
+  return ((double)t1->sec - t0->sec)*1000 + ((ngx_int_t)(t1->msec) - (ngx_int_t)(t0->msec));
+}
+
+static void nchan_accumulator_rebase_to_time(nchan_accumulator_t *acc, ngx_time_t *t1) {
+  double         diff   = ngx_time_t_diff(&acc->last_update, t1);
+  double         pow_mult;
+  
+  if(diff == 0) {
+    return;
+  }
+  
+  acc->last_update = *t1;
+  switch(acc->type) {
+    case ACCUMULATOR_EXP_DECAY_FLOAT:
+      //the accurate way
+      pow_mult = pow(2, -1 * acc->data.ed_float.lambda * diff);
+      acc->data.ed_float.value = acc->data.ed_float.value * pow_mult;
+      acc->data.ed_float.weight= acc->data.ed_float.weight * pow_mult;
+      break;
+    default:
+      //not supported yet
+      break;
+  }
+}
+
+static void nchan_accumulator_rebase_to_now(nchan_accumulator_t *acc) {
+  ngx_time_t    *t1     = ngx_timeofday();
+  nchan_accumulator_rebase_to_time(acc, t1);
+}
+
+
+int nchan_accumulator_update(nchan_accumulator_t *acc, double val) {
+  switch(acc->type) {
+    case ACCUMULATOR_EXP_DECAY_FLOAT:
+      //the accurate way
+      nchan_accumulator_rebase_to_now(acc);
+      acc->data.ed_float.value += val;
+      acc->data.ed_float.weight += 1;
+      break;
+    case ACCUMULATOR_SUM:
+      acc->data.sum.value += val;
+      acc->data.sum.weight += 1;
+      break;
+    default:
+      //not supported yet
+      return 0;
+  }
+  return 1;
+}
+
+int nchan_accumulator_atomic_update(nchan_accumulator_t *acc, double val) {
+  
+  int intval;
+  switch(acc->type) {
+    case ACCUMULATOR_EXP_DECAY_FLOAT:
```

Comment:
Can we mere this into:
```
case ACCUMULATOR_EXP_DECAY_FLOAT: //not possible for now
default:
      //not supported yet
      return 0;
```

Reply:
Could, but won't -- I like keeping these cases separate and in the same order as in the `enum`